#+TITLE: Meta-Barcoding Dual-Indexing
#+AUTHOR: Markus Ankenbrand

* About
This is a collection of the computational methods used for the publication <Link will be included when available>.
Here we provide all custom scripts and commands to replicate our results.
Furthermore you can download the pre-computed training sets for the RDPclassifier <Link> and UTAX <Link>.
If you find this information useful please consider citing our article <Link>.

* Data
If you are just interested in the pre-computed training sets go to releases and download the tar-ball.
There you can find the training sets in the subfolder training.

* Dependencies
This is a list of software tools and versions used in the original analysis.
We provide precomputed results for most steps so don't worry if you can't get a certain tool in a certain version.
If you get substantially different results (e.g. using other versions/tools) feel free to contact me.
You can also do so by opening an Issue here on GitHub.
** Hardware
Most of the analyses were performed on a Laptop with
 - Intel(R) Core(TM) i7-4800MQ CPU @ 2.70GHz Processor
 - 16GB RAM
 - Ubuntu 14.04 operating system
Only the RDP training was performed on a machine with more RAM available.
** Third party tools
 - [[http://its2.bioapps.biozentrum.uni-wuerzburg.de/][ITS2 database]] (Version 4.0.0 and inofficial release (included))
 - [[http://www.drive5.com/usearch/download.html][UTAX]] (Version usearch v8.0.1477_i86linux32)
 - [[https://github.com/rdpstaff/classifier][RDPclassifier]] (Version master from GitHub (3ba79f222f7a445620549633e603436c2f0eb7ba, 2015-02-20))
 - [[https://github.com/greatfireball/NCBI-Taxonomy/tree/v0.70.5][NCBI::Taxonomy]] - perl module (Version v0.70.5, [[http://dx.doi.org/10.5281/zenodo.17375][doi:10.5281/zenodo.17375]])
 - [[http://www.r-project.org/][R]] (Version 3.1.2 (2014-10-31) -- "Pumpkin Helmet", Platform: x86_64-pc-linux-gnu (64-bit))
 - [[https://www.perl.org/][perl]] (Version v5.20.2 built for x86_64-linux-gnu-thread-multi)

* Workflow
** Preparation of Reference Database
*** Data
The ITS2 sequences can be retrieved from the ITS2 database.
The previous reference db consisted of all "Direct folds & Homology modeled" sequences from "Viridiplantae" (db version v4.0.0, 2011).
This [[data/viridiplantae_folds_2011.fasta][file]] is included for comparison.
The new reference database consists of all sequences from "Viridiplantae" of an inofficial ITS2 database release (2014).
This [[data/viridiplantae_all_2014.fasta][file]] is also included in this repository.
*** Taxonomic assignment
Assign NCBI taxonomy on kingdom, phylum, class, order, family, genus and species level by mapping gi to taxid.
Create an analysis folder and execute the following commands there (you need the NCBI::Taxonomy module for this):
#+BEGIN_SRC bash :dir analysis
grep "^>" ../data/viridiplantae_all_2014.fasta |
 perl -pe 's/^>(\d+).*/$1/' >viridiplantae_all_2014.gis
perl ../code/gi2taxonomy.pl\
 --gis viridiplantae_all_2014.gis\
 --out viridiplantae_all_2014.tax\
 --species viridiplantae_all_2014.species.taxids\
 --genus viridiplantae_all_2014.genus.taxids 
#+END_SRC
This generates the following files:
 - viridiplantae_all_2014.gis
 - viridiplantae_all_2014.tax
 - viridiplantae_all_2014.species.taxids
 - viridiplantae_all_2014.genus.taxids
All of those are also included in the precomputed folder.
*** UTAX and RDP training
The following commands executed in the analysis folder generate the required fasta and tax files for RDP and UTAX:
#+BEGIN_SRC bash :dir analysis
perl ../code/tax2rdp_utax.pl viridiplantae_all_2014.tax\
 ../data/viridiplantae_all_2014.fasta viridiplantae_all_2014
#+END_SRC
This generates the following files:
 - viridiplantae_all_2014.gi_tax.map
 - viridiplantae_all_2014.rdp.fa
 - viridiplantae_all_2014.rdp.tax
 - viridiplantae_all_2014.utax.fa
 - viridiplantae_all_2014.utax.tax
The first three are also included in the precomputed folder. And the last two are included in the training/utax folder.
The utax files are ready to be used for classification. 
However to speed up the initial step a udb file can be created as follows:
#+BEGIN_SRC bash :dir analysis
usearch8 -makeudb_usearch viridiplantae_all_2014.utax.fa\
 -output viridiplantae_all_2014.utax.udb
#+END_SRC
This creates the file viridiplantae_all_2014.utax.udb which is not included as it is not required and its size is 225MB.
To train the RDPclassifier execute the following commands 
(warning for the train command 16GB RAM did not suffice, but 32 did):
#+BEGIN_SRC bash :dir analysis
mkdir rdp_trained

java -jar classifier.jar rm-dupseq --infile viridiplantae_all_2014.rdp.fa\
 --outfile viridiplantae_all_2014.rdp.rm-dupseq.fa\
 --duplicates --min_seq_length 150

java -jar classifier.jar rm-partialseq viridiplantae_all_2014.rdp.fa\
 viridiplantae_all_2014.rdp.rm-dupseq.fa\
 viridiplantae_all_2014.rdp.rm-dupseq.rm-partialseq.fa\
 --alignment-mode overlap --min_gaps 50 --knn 20

java -Xmx32g -jar classifier.jar train --out_dir rdp_trained\
 --seq viridiplantae_all.rdp.rm-dupseq.rm-partialseq.fa\
 --tax_file viridiplantae_all.rdp.tax

cp data/its2.properties rdp_trained/its2.properties
#+END_SRC
This generates the following files:
 - viridiplantae_all_2014.rdp.rm-dupseq.fa
 - viridiplantae_all_2014.rdp.rm-dupseq.rm-partialseq.fa
All of those are also included in the precomputed folder.
And the folder rdp_trained including five files:
 - rdp_trained/bergeyTrainingTree.xml
 - rdp_trained/genus_wordConditionalProbList.txt
 - rdp_trained/its2.properties
 - rdp_trained/wordConditionalProbIndexArr.txt
 - rdp_trained/logWordPrior.txt
Those are the files required for RDP classification and are included as a tar.gz file in training/rdp
*** Species coverage
** Analysis of Pollen Samples
*** Data
Download data from [[http://www.ebi.ac.uk/][EBI]] SRA repository project accession number PRJEB8640.
*** Preprocessing
*** Classification
*** Species accumulation curves
